lib/mix/lib/mix/compilers/erlang.ex

```diff
@@ -188,16 +190,60 @@ defmodule Mix.Compilers.Erlang do
     artifact |> Path.basename |> Path.rootname
   end
 
+  # The manifest file contains a list of {dest, warnings} tuples
   defp read_manifest(file) do
-    case File.read(file) do
-      {:ok, contents} -> String.split(contents, "\n")
-      {:error, _} -> []
+    try do
+      file |> File.read!() |> :erlang.binary_to_term()
+    rescue
+      _ -> []
+    else
+      {@manifest_vsn, data} when is_list(data) -> data
+      _ -> []
     end
   end
 
   defp write_manifest(file, entries, timestamp) do
     Path.dirname(file) |> File.mkdir_p!
-    File.write!(file, Enum.join(entries, "\n"))
+    File.write!(file, :erlang.term_to_binary({@manifest_vsn, entries}))
     File.touch!(file, timestamp)
   end
+
+  defp do_compile({input, output}, callback, timestamp, verbose) do
+    case callback.(input, output) do
```

Comment:
This is backwards incompatible. We need to still accept `{:ok, _}` and we need to have a catch all `_` at the end too. For the catch all, we can emit a deprecation warning now:

    other ->
      IO.warn "callback given to #{inspect __MODULE__} returned #{inspect other}. " <>
              "Expected {:ok, term, warnings} or {:error, errors, warnings}"

Reply:
Actually, the advice above is wrong, since we also supported `:error`. So we still need to accept `{:ok, _}` and `:error`, we don't need a catch all. We can't deprecate anything for now but you should add a TODO like this:

```elixir
# TODO: Deprecate {:ok, _} and :error return on Elixir v1.8
```